hw/ip/otbn/dv/rig/rig/gens/known_wdr.py

```diff
@@ -0,0 +1,109 @@
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+
+import random
+from typing import Optional
+
+from shared.operand import ImmOperandType, RegOperandType
+from shared.insn_yaml import InsnsFile
+
+from ..config import Config
+from ..model import Model
+from ..program import ProgInsn, Program
+
+from ..snippet import ProgSnippet
+from ..snippet_gen import GenCont, GenRet, SnippetGen
+
+
+class KnownWDR(SnippetGen):
+    '''A snippet generator that generates known values (all zeros or all ones
+    for now) for WDRs.
+
+    '''
+
+    def __init__(self, cfg: Config, insns_file: InsnsFile) -> None:
+        super().__init__()
+
+        if 'bn.xor' not in insns_file.mnemonic_to_insn:
+            raise RuntimeError('BN.XOR instruction not in instructions file')
+
+        if 'bn.not' not in insns_file.mnemonic_to_insn:
+            raise RuntimeError('BN.NOT instruction not in instructions file')
+
+        self.insn_xor = insns_file.mnemonic_to_insn['bn.xor']
+        self.insn_not = insns_file.mnemonic_to_insn['bn.not']
+        # BN.XOR has six operands: wrd, wrs1, wrs2, shift_type, shift_value
+        # and flag_group
+        if not (len(self.insn_xor.operands) == 6 and
+                isinstance(self.insn_xor.operands[0].op_type, RegOperandType) and
+                self.insn_xor.operands[0].op_type.reg_type == 'wdr' and
+                self.insn_xor.operands[0].op_type.is_dest() and
+                isinstance(self.insn_xor.operands[1].op_type, RegOperandType) and
+                self.insn_xor.operands[1].op_type.reg_type == 'wdr' and
+                not self.insn_xor.operands[1].op_type.is_dest() and
+                isinstance(self.insn_xor.operands[2].op_type, RegOperandType) and
+                self.insn_xor.operands[2].op_type.reg_type == 'wdr' and
+                not self.insn_xor.operands[2].op_type.is_dest() and
+                isinstance(self.insn_xor.operands[4].op_type, ImmOperandType)):
+            raise RuntimeError('BN.XOR instruction from instructions file is '
+                               'not the shape expected by the SmallVal generator.')
+
+        self.wrd_op_type = self.insn_xor.operands[0].op_type
+        self.wrs1_op_type = self.insn_xor.operands[1].op_type
+        self.imm_op_type = self.insn_xor.operands[4].op_type
+        self.imm_op_type.shift = 3
+
+    def gen(self,
+            cont: GenCont,
+            model: Model,
+            program: Program) -> Optional[GenRet]:
+        # Return None if this is the last instruction in the current gap
```

Comment:
Nit: This needs re-wording slightly because we're checking for `<= 3`, not `<= 1`.